In [ ]:
import os
import sys

current_dir = os.getcwd()

# Walk up the directory tree until we find 'src'
path = current_dir
src_path = None

while True:
    if os.path.basename(path) == "src":
        src_path = path
        break
    parent = os.path.dirname(path)
    if parent == path:  # reached filesystem root
        break
    path = parent

# Add src to sys.path if found
if src_path and src_path not in sys.path:
    sys.path.insert(0, src_path)

import os
import sys
import time
import numpy as np
import pandas as pd
import scipy.sparse as sp
import torch
import torch.nn as nn
import torch.optim as optim
import torch.utils.data as data
import torch.backends.cudnn as cudnn
import torch.nn.functional as F

# Add the src directory to Python path to enable imports from helpers
# The notebook is in src/ncf/, so we go up one level to get to src/
current_dir = os.getcwd()
# Navigate to src directory (parent of ncf directory)
if os.path.basename(current_dir) == 'ncf':
    src_path = os.path.dirname(current_dir)
else:
    # Try to find src in the current path
    parts = current_dir.split(os.sep)
    if 'src' in parts:
        src_idx = parts.index('src')
        src_path = os.sep.join(parts[:src_idx + 1])
    else:
        # Fallback: assume src is a sibling or child directory
        src_path = os.path.join(os.path.dirname(current_dir), 'src')

if os.path.exists(src_path) and src_path not in sys.path:
    sys.path.insert(0, src_path)

# Local imports
from helpers import download_ml1m_dataset
from utils.ml_to_ncf import preprocess_ml1m_to_ncf_format
from utils.ncfdata import NCFData
from helpers.ncf_model import NCF


In [7]:
dataset = 'ml-1m'
# ============================================================================
# MODEL ARCHITECTURE CONFIGURATION
# ============================================================================

# Choose which model architecture to use
# Options:
#   - 'MLP': Multi-Layer Perceptron only (non-linear interactions)
#   - 'GMF': Generalized Matrix Factorization only (linear interactions)
#   - 'NeuMF-end': Neural Matrix Factorization trained from scratch (end-to-end)
#   - 'NeuMF-pre': Neural Matrix Factorization with pre-trained GMF and MLP models
model_name = 'NeuMF-end'
assert model_name in ['MLP', 'GMF', 'NeuMF-end', 'NeuMF-pre'], \
    f"Model must be 'MLP', 'GMF', 'NeuMF-end', or 'NeuMF-pre', got '{model_name}'"

print(f"✓ Model architecture: {model_name}")

# ============================================================================
# DATA AND MODEL PATHS CONFIGURATION
# ============================================================================

# Data will be downloaded automatically during training
# We'll create a local data directory to store downloaded files

data_dir = os.path.join(os.path.dirname(os.getcwd()), '..', 'data')
if not os.path.exists(data_dir):
    os.makedirs(data_dir)

# Model saving directory
model_path = os.path.join(os.path.dirname(os.getcwd()), '..', 'models')
if not os.path.exists(model_path):
    os.makedirs(model_path)

GMF_model_path = os.path.join(model_path, 'GMF.pth')
MLP_model_path = os.path.join(model_path, 'MLP.pth')
NeuMF_model_path = os.path.join(model_path, 'NeuMF.pth')

print(f"✓ Directories configured")
print(f"  - Data directory: {data_dir} (will be created/used for downloaded data)")
print(f"  - Model save path: {model_path}")

# ============================================================================
# 2.4 TRAINING HYPERPARAMETERS
# ============================================================================

# Learning rate: Controls how big steps the optimizer takes during training
# Too high: training might be unstable or diverge
# Too low: training will be very slow
# Typical range: 0.0001 to 0.01
learning_rate = 0.001

# Dropout rate: Regularization technique to prevent overfitting
# Randomly sets some neurons to zero during training
# Range: 0.0 (no dropout) to 0.9 (very aggressive dropout)
# 0.0 means no dropout (all neurons active)
dropout_rate = 0.1

# Batch size: Number of training examples processed together in one iteration
# Larger batch size: more stable gradients, but requires more memory
# Smaller batch size: less memory, but noisier gradients
# Typical values: 64, 128, 256, 512
batch_size = 256

# Number of training epochs: How many times we'll iterate through the entire dataset
# More epochs: better learning, but risk of overfitting
# Too few epochs: model might not learn enough
epochs = 20

# Top-K for evaluation: When evaluating, we recommend top K items to each user
# We measure if the true item is in the top K recommendations
# Common values: 5, 10, 20
top_k = 10

# Factor number: Dimension of the embedding vectors for users and items
# Larger: more capacity to learn complex patterns, but more parameters
# Smaller: fewer parameters, faster training, but less capacity
# Common values: 8, 16, 32, 64
factor_num = 32

# Number of MLP layers: Depth of the Multi-Layer Perceptron component
# More layers: can learn more complex non-linear patterns
# Fewer layers: simpler model, faster training
# Typical range: 1 to 5 layers
num_layers = 3

# Number of negative samples for training: For each positive (user, item) pair,
# we sample this many negative items (items the user hasn't interacted with)
# More negatives: better learning signal, but slower training
# Fewer negatives: faster training, but potentially weaker learning
# Common values: 1, 4, 8
num_ng = 4

# Number of negative samples for testing: During evaluation, for each test item,
# we also provide this many negative items. The model should rank the true item higher.
# Typically 99 negatives + 1 positive = 100 items total per test case
test_num_ng = 99

# Whether to save the trained model
save_model = True

✓ Model architecture: NeuMF-end
✓ Directories configured
  - Data directory: /Users/abbas/Documents/Codes/thesis/recommender/src/../data (will be created/used for downloaded data)
  - Model save path: /Users/abbas/Documents/Codes/thesis/recommender/src/../models


In [23]:
print("=" * 70)
print("STEP 3.1: Downloading MovieLens 1M Dataset")
print("=" * 70)
ratings_file = download_ml1m_dataset(data_dir)


STEP 3.1: Downloading MovieLens 1M Dataset
✓ Dataset already exists at /Users/abbas/Documents/Codes/thesis/recommender/src/../data/ml-1m/ratings.dat


In [24]:
train_rating_path, test_rating_path, test_negative_path, user_num, item_num, train_mat = \
    preprocess_ml1m_to_ncf_format(ratings_file, data_dir, test_ratio=0.2, test_negatives=99)


✓ Loaded 1000209 ratings
  - Unique users: 6040
  - Unique movies: 3706

Filtering positive interactions (ratings >= 4)...
✓ 575281 positive interactions (out of 1000209 total)

Remapping user and item IDs to be contiguous...
✓ Remapped to 6038 users and 3533 items

Splitting data (train: 80%, test: 20%)...
✓ Training pairs: 460225
✓ Test pairs: 115056

Saving training data to /Users/abbas/Documents/Codes/thesis/recommender/src/../data/ml-1m.train.rating...
✓ Saved 460225 training pairs

Creating training interaction matrix...
✓ Training matrix created: 460225 interactions

Generating test negative samples (99 negatives per test case)...
✓ Generated test negative samples: 115056 test cases


In [25]:
# ============================================================================
# STEP 4: PYTORCH DATASET CLASS AND DATA LOADING
# ============================================================================

"""
This step creates the PyTorch Dataset class and data loading functions.
"""

# ============================================================================
# 4.1 LOAD ALL DATA FILES
# ============================================================================

def load_all_data(train_rating_path, test_negative_path):
    """
    Loads all data files into memory for efficient access during training.
    
    This function loads:
    1. Training data: user-item pairs from train.rating file
    2. Test data: user-item pairs with negative samples from test.negative file
    3. Training matrix: sparse matrix for efficient negative sampling
    
    Parameters:
    - train_rating_path: Path to the training rating file
    - test_negative_path: Path to the test negative file
    
    Returns:
    - train_data: List of [user, item] pairs for training
    - test_data: List of [user, item] pairs for testing (includes negatives)
    - user_num: Total number of users
    - item_num: Total number of items
    - train_mat: Sparse matrix of training interactions
    """
    print("=" * 70)
    print("STEP 4.1: Loading Data Files")
    print("=" * 70)
    
    # Load training data
    print(f"Loading training data from {train_rating_path}...")
    train_data = pd.read_csv(
        train_rating_path,
        sep='\t',
        header=None,
        names=['user', 'item'],
        usecols=[0, 1],
        dtype={0: np.int32, 1: np.int32}
    )
    
    # Calculate number of users and items
    user_num = train_data['user'].max() + 1
    item_num = train_data['item'].max() + 1
    
    print(f"✓ Loaded {len(train_data)} training pairs")
    print(f"  - Users: {user_num}")
    print(f"  - Items: {item_num}")
    
    # Convert to list of lists for easier processing
    train_data = train_data.values.tolist()
    
    # Create sparse training matrix (Dictionary of Keys format)
    # This is used to quickly check if a user-item pair exists in training data
    print("\nCreating training interaction matrix...")
    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.float32)
    for u, i in train_data:
        train_mat[u, i] = 1.0
    print(f"✓ Training matrix created: {train_mat.nnz} interactions")
    
    # Load test data with negative samples
    print(f"\nLoading test data from {test_negative_path}...")
    test_data = []
    with open(test_negative_path, 'r') as fd:
        line = fd.readline()
        while line is not None and line != '':
            # Format: (user, item)\tneg1\tneg2\t...\tneg99
            arr = line.strip().split('\t')
            
            # Parse the positive pair: (user, item)
            # eval() converts string "(123, 456)" to tuple (123, 456)
            positive_pair = eval(arr[0])
            u = positive_pair[0]
            i = positive_pair[1]
            
            # Add the positive pair
            test_data.append([u, i])
            
            # Add all negative items for this user
            for neg_item in arr[1:]:
                if neg_item:  # Skip empty strings
                    test_data.append([u, int(neg_item)])
            
            line = fd.readline()
    
    print(f"✓ Loaded {len(test_data)} test pairs (including negatives)")
    
    print("\n" + "=" * 70)
    print("✓ Data loading complete!")
    print("=" * 70)
    
    return train_data, test_data, user_num, item_num, train_mat

# Load all data
train_data, test_data, user_num, item_num, train_mat = load_all_data(
    train_rating_path, test_negative_path
)



STEP 4.1: Loading Data Files
Loading training data from /Users/abbas/Documents/Codes/thesis/recommender/src/../data/ml-1m.train.rating...
✓ Loaded 460225 training pairs
  - Users: 6038
  - Items: 3533

Creating training interaction matrix...
✓ Training matrix created: 460225 interactions

Loading test data from /Users/abbas/Documents/Codes/thesis/recommender/src/../data/ml-1m.test.negative...
✓ Loaded 11505600 test pairs (including negatives)

✓ Data loading complete!


In [26]:
train_dataset = NCFData(
    train_data,
    item_num,
    train_mat,
    num_ng=num_ng,  # From Step 2 configuration
    is_training=True
)

test_dataset = NCFData(
    test_data,
    item_num,
    train_mat,
    num_ng=0,  # No negative sampling for testing
    is_training=False
)

train_loader = data.DataLoader(
    train_dataset,
    batch_size=batch_size,  # From Step 2 configuration
    shuffle=True,  # Shuffle training data each epoch
    num_workers=0,  # MUST be 0 for Jupyter notebooks (avoids pickling errors)
    pin_memory=True if torch.cuda.is_available() else False  # Faster GPU transfer
)

test_loader = data.DataLoader(
    test_dataset,
    batch_size=test_num_ng + 1,  # 1 positive + test_num_ng negatives
    shuffle=False,  # Don't shuffle test data
    num_workers=0,  # MUST be 0 for Jupyter notebooks
    pin_memory=True if torch.cuda.is_available() else False
)

# Verify num_workers is 0 (safety check)
assert train_loader.num_workers == 0, f"ERROR: train_loader.num_workers is {train_loader.num_workers}, must be 0!"
assert test_loader.num_workers == 0, f"ERROR: test_loader.num_workers is {test_loader.num_workers}, must be 0!"

In [27]:
# ============================================================================
# 5.2 CREATE AND INITIALIZE THE MODEL
# ============================================================================

print("\n" + "=" * 70)
print("STEP 5.2: Creating and Initializing Model")
print("=" * 70)

# Check if we need pre-trained models (for NeuMF-pre)
if model_name == 'NeuMF-pre':
    # For NeuMF-pre, we would load pre-trained GMF and MLP models
    # For now, we'll use NeuMF-end (training from scratch)
    print("⚠ NeuMF-pre requires pre-trained models.")
    print("  Switching to NeuMF-end (training from scratch)...")
    model_name = 'NeuMF-end'

# Create the model
ncf_model = NCF(
    user_num=user_num,
    item_num=item_num,
    factor_num=factor_num,
    num_layers=num_layers,
    dropout=dropout_rate,
    model_name=model_name,
    GMF_model=None,
    MLP_model=None
)

# Move model to GPU if available
if torch.cuda.is_available():
    ncf_model = ncf_model.cuda()
    print("✓ Model moved to GPU")
else:
    print("✓ Model on CPU")

# Count parameters
total_params = sum(p.numel() for p in ncf_model.parameters())
trainable_params = sum(p.numel() for p in ncf_model.parameters() if p.requires_grad)

print(f"\n✓ Model created successfully!")
print(f"  - Total parameters: {total_params:,}")
print(f"  - Trainable parameters: {trainable_params:,}")

# Print model architecture
print(f"\nModel Architecture:")
print(f"  - Users: {user_num:,}")
print(f"  - Items: {item_num:,}")
if model_name != 'MLP':
    print(f"  - GMF embeddings: {factor_num} dimensions")
if model_name != 'GMF':
    mlp_embed_dim = factor_num * (2 ** (num_layers - 1))
    print(f"  - MLP embeddings: {mlp_embed_dim} dimensions")
    print(f"  - MLP layers: {num_layers} (with dropout={dropout_rate})")
print(f"  - Prediction layer: {factor_num if model_name in ['MLP', 'GMF'] else factor_num * 2} → 1")



# Print the full model structure
print("\n" + "=" * 70)
print("COMPLETE MODEL STRUCTURE:")
print("=" * 70)
print(ncf_model)
print("=" * 70)

# Print detailed layer information
print("\n" + "=" * 70)
print("LAYER-BY-LAYER BREAKDOWN:")
print("=" * 70)

if model_name != 'MLP':
    print("\n[GMF Path - Generalized Matrix Factorization]")
    print(f"  embed_user_GMF: Embedding({user_num}, {factor_num})")
    print(f"    → Converts user IDs to {factor_num}-dimensional vectors")
    print(f"  embed_item_GMF: Embedding({item_num}, {factor_num})")
    print(f"    → Converts item IDs to {factor_num}-dimensional vectors")
    print(f"  Element-wise product: user_emb * item_emb")
    print(f"    → Output shape: [batch_size, {factor_num}]")

if model_name != 'GMF':
    mlp_embed_dim = factor_num * (2 ** (num_layers - 1))
    print("\n[MLP Path - Multi-Layer Perceptron]")
    print(f"  embed_user_MLP: Embedding({user_num}, {mlp_embed_dim})")
    print(f"    → Converts user IDs to {mlp_embed_dim}-dimensional vectors")
    print(f"  embed_item_MLP: Embedding({item_num}, {mlp_embed_dim})")
    print(f"    → Converts item IDs to {mlp_embed_dim}-dimensional vectors")
    print(f"  Concatenation: [user_emb, item_emb]")
    print(f"    → Output shape: [batch_size, {mlp_embed_dim * 2}]")
    
    print(f"\n  MLP Layers ({num_layers} layers):")
    for i in range(num_layers):
        input_size = factor_num * (2 ** (num_layers - i))
        output_size = input_size // 2
        print(f"    Layer {i+1}:")
        print(f"      Dropout(p={dropout_rate})")
        print(f"      Linear({input_size}, {output_size})")
        print(f"      ReLU()")
        print(f"      → Output shape: [batch_size, {output_size}]")

print("\n[Prediction Layer]")
if model_name == 'GMF':
    predict_input = factor_num
    print(f"  Input: GMF output [{factor_num} dimensions]")
elif model_name == 'MLP':
    predict_input = factor_num
    print(f"  Input: MLP output [{factor_num} dimensions]")
else:  # NeuMF
    predict_input = factor_num * 2
    print(f"  Input: Concatenated GMF + MLP [{factor_num * 2} dimensions]")
    print(f"    → GMF: [{factor_num} dims] + MLP: [{factor_num} dims]")

print(f"  Linear({predict_input}, 1)")
print(f"    → Output: [batch_size, 1] (interaction score)")
print(f"    → Higher score = more likely user will like item")

# Calculate and print parameter breakdown
print("\n" + "=" * 70)
print("PARAMETER BREAKDOWN:")
print("=" * 70)

total_params = 0
if model_name != 'MLP':
    gmf_user_params = user_num * factor_num
    gmf_item_params = item_num * factor_num
    gmf_total = gmf_user_params + gmf_item_params
    total_params += gmf_total
    print(f"\nGMF Embeddings:")
    print(f"  User embeddings: {user_num:,} × {factor_num} = {gmf_user_params:,} parameters")
    print(f"  Item embeddings: {item_num:,} × {factor_num} = {gmf_item_params:,} parameters")
    print(f"  GMF Total: {gmf_total:,} parameters")

if model_name != 'GMF':
    mlp_embed_dim = factor_num * (2 ** (num_layers - 1))
    mlp_user_params = user_num * mlp_embed_dim
    mlp_item_params = item_num * mlp_embed_dim
    mlp_embed_total = mlp_user_params + mlp_item_params
    total_params += mlp_embed_total
    print(f"\nMLP Embeddings:")
    print(f"  User embeddings: {user_num:,} × {mlp_embed_dim} = {mlp_user_params:,} parameters")
    print(f"  Item embeddings: {item_num:,} × {mlp_embed_dim} = {mlp_item_params:,} parameters")
    print(f"  MLP Embeddings Total: {mlp_embed_total:,} parameters")
    
    # MLP layers parameters
    mlp_layer_params = 0
    print(f"\nMLP Layers:")
    for i in range(num_layers):
        input_size = factor_num * (2 ** (num_layers - i))
        output_size = input_size // 2
        layer_params = (input_size * output_size) + output_size  # weights + bias
        mlp_layer_params += layer_params
        print(f"  Layer {i+1} (Linear({input_size}, {output_size})): {layer_params:,} parameters")
    total_params += mlp_layer_params
    print(f"  MLP Layers Total: {mlp_layer_params:,} parameters")

# Prediction layer
if model_name in ['MLP', 'GMF']:
    predict_input = factor_num
else:
    predict_input = factor_num * 2
predict_params = (predict_input * 1) + 1  # weights + bias
total_params += predict_params
print(f"\nPrediction Layer:")
print(f"  Linear({predict_input}, 1): {predict_params:,} parameters")

print("\n" + "=" * 70)
print(f"TOTAL MODEL PARAMETERS: {total_params:,}")
print(f"Model Size (float32): ~{total_params * 4 / 1024 / 1024:.2f} MB")
print("=" * 70)


STEP 5.2: Creating and Initializing Model
✓ Model on CPU

✓ Model created successfully!
  - Total parameters: 1,574,657
  - Trainable parameters: 1,574,657

Model Architecture:
  - Users: 6,038
  - Items: 3,533
  - GMF embeddings: 32 dimensions
  - MLP embeddings: 128 dimensions
  - MLP layers: 3 (with dropout=0.1)
  - Prediction layer: 64 → 1

COMPLETE MODEL STRUCTURE:
NCF(
  (embed_user_GMF): Embedding(6038, 32)
  (embed_item_GMF): Embedding(3533, 32)
  (embed_user_MLP): Embedding(6038, 128)
  (embed_item_MLP): Embedding(3533, 128)
  (MLP_layers): Sequential(
    (0): Dropout(p=0.1, inplace=False)
    (1): Linear(in_features=256, out_features=128, bias=True)
    (2): ReLU()
    (3): Dropout(p=0.1, inplace=False)
    (4): Linear(in_features=128, out_features=64, bias=True)
    (5): ReLU()
    (6): Dropout(p=0.1, inplace=False)
    (7): Linear(in_features=64, out_features=32, bias=True)
    (8): ReLU()
  )
  (predict_layer): Linear(in_features=64, out_features=1, bias=True)
)

LAYER-B

In [10]:
# ============================================================================
# STEP 6: EVALUATION METRICS
# ============================================================================

"""
This step implements evaluation metrics for recommendation systems:
- Hit Rate (HR@K): Binary metric - is the true item in top K?
- NDCG (Normalized Discounted Cumulative Gain@K): Ranking quality metric
"""

# ============================================================================
# 6.1 HIT RATE METRIC
# ============================================================================

def hit(gt_item, pred_items):
    """
    Calculate Hit Rate for a single test case.
    
    Hit Rate is 1 if the ground truth item is in the predicted top-K items,
    otherwise 0.
    
    Parameters:
    - gt_item: Ground truth item ID (the item user actually interacted with)
    - pred_items: List of top-K predicted item IDs (recommended items)
    
    Returns:
    - 1 if gt_item is in pred_items, 0 otherwise
    """
    if gt_item in pred_items:
        return 1
    return 0

print("=" * 70)
print("STEP 6.1: Hit Rate Metric")
print("=" * 70)
print("✓ Hit Rate function defined")
print("  - Returns 1 if true item is in top-K recommendations")
print("  - Returns 0 otherwise")


STEP 6.1: Hit Rate Metric
✓ Hit Rate function defined
  - Returns 1 if true item is in top-K recommendations
  - Returns 0 otherwise


In [11]:
# ============================================================================
# 6.2 NDCG METRIC
# ============================================================================

def ndcg(gt_item, pred_items):
    """
    Calculate Normalized Discounted Cumulative Gain (NDCG) for a single test case.
    
    NDCG measures ranking quality by:
    1. Giving more weight to items ranked higher (position matters)
    2. Using logarithmic discounting (relevance decreases with position)
    
    Formula: NDCG = 1 / log2(position + 2)
    - Position 0 (top): 1 / log2(2) = 1.0
    - Position 1: 1 / log2(3) ≈ 0.63
    - Position 2: 1 / log2(4) = 0.5
    - Position 9: 1 / log2(11) ≈ 0.29
    
    Parameters:
    - gt_item: Ground truth item ID (the item user actually interacted with)
    - pred_items: List of top-K predicted item IDs (recommended items)
    
    Returns:
    - NDCG score (0.0 to 1.0) if gt_item is in pred_items
    - 0.0 if gt_item is not in pred_items
    """
    if gt_item in pred_items:
        # Find the position (index) of the ground truth item
        index = pred_items.index(gt_item)
        # Calculate NDCG: 1 / log2(position + 2)
        # +2 because: position 0 should give 1/log2(2) = 1.0
        return np.reciprocal(np.log2(index + 2))
    return 0.0

print("=" * 70)
print("STEP 6.2: NDCG Metric")
print("=" * 70)
print("✓ NDCG function defined")
print("  - Measures ranking quality")
print("  - Higher score for items ranked higher")
print("  - Returns 0 if true item not in recommendations")

# Example to demonstrate NDCG
print("\nNDCG Examples:")
example_items = [10, 20, 30, 40, 50]
print(f"  Top-5 recommendations: {example_items}")
print(f"  If true item is at position 0: NDCG = {ndcg(10, example_items):.3f}")
print(f"  If true item is at position 2: NDCG = {ndcg(30, example_items):.3f}")
print(f"  If true item is at position 4: NDCG = {ndcg(50, example_items):.3f}")
print(f"  If true item not in list: NDCG = {ndcg(99, example_items):.3f}")


STEP 6.2: NDCG Metric
✓ NDCG function defined
  - Measures ranking quality
  - Higher score for items ranked higher
  - Returns 0 if true item not in recommendations

NDCG Examples:
  Top-5 recommendations: [10, 20, 30, 40, 50]
  If true item is at position 0: NDCG = 1.000
  If true item is at position 2: NDCG = 0.500
  If true item is at position 4: NDCG = 0.387
  If true item not in list: NDCG = 0.000


In [12]:
# ============================================================================
# 6.3 EVALUATION FUNCTION
# ============================================================================

def evaluate_metrics(model, test_loader, top_k, device='cuda'):
    """
    Evaluate model performance on test data.
    
    This function:
    1. For each test case (1 positive + 99 negatives):
       - Gets model predictions for all 100 items
       - Selects top-K items with highest scores
       - Checks if the true item is in top-K (Hit Rate)
       - Calculates NDCG based on true item's position
    2. Averages metrics across all test cases
    
    Parameters:
    - model: Trained NCF model
    - test_loader: DataLoader with test data
    - top_k: Number of top items to consider (e.g., 10)
    - device: 'cuda' or 'cpu'
    
    Returns:
    - mean_HR: Average Hit Rate across all test cases
    - mean_NDCG: Average NDCG across all test cases
    """
    model.eval()  # Set model to evaluation mode (disables dropout)
    
    HR_list = []  # List to store Hit Rate for each test case
    NDCG_list = []  # List to store NDCG for each test case
    
    with torch.no_grad():  # Disable gradient computation (faster, saves memory)
        for user, item, label in test_loader:
            # Move data to device (GPU or CPU)
            if device == 'cuda' and torch.cuda.is_available():
                user = user.cuda()
                item = item.cuda()
            else:
                device = 'cpu'
            
            # Get model predictions for all items in this batch
            # Batch size = test_num_ng + 1 = 100 (1 positive + 99 negatives)
            predictions = model(user, item)  # [100] tensor of scores
            
            # Get top-K items with highest prediction scores
            # torch.topk returns (values, indices)
            _, indices = torch.topk(predictions, top_k)
            
            # Get the actual item IDs for top-K recommendations
            # torch.take extracts items at given indices
            recommends = torch.take(item, indices).cpu().numpy().tolist()
            
            # The first item in the batch is always the positive (true) item
            gt_item = item[0].item()  # Ground truth item ID
            
            # Calculate metrics for this test case
            HR_list.append(hit(gt_item, recommends))
            NDCG_list.append(ndcg(gt_item, recommends))
    
    # Calculate average metrics across all test cases
    mean_HR = np.mean(HR_list)
    mean_NDCG = np.mean(NDCG_list)
    
    return mean_HR, mean_NDCG

print("=" * 70)
print("STEP 6.3: Evaluation Function")
print("=" * 70)
print("✓ Evaluation function defined")
print("  - Evaluates model on test data")
print("  - Calculates average Hit Rate and NDCG")
print("  - Works with GPU or CPU")

# Determine device
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f"\n✓ Evaluation ready (device: {device})")


STEP 6.3: Evaluation Function
✓ Evaluation function defined
  - Evaluates model on test data
  - Calculates average Hit Rate and NDCG
  - Works with GPU or CPU

✓ Evaluation ready (device: cpu)


### Detailed Explanation of Step 6:

#### 6.1 Hit Rate (HR@K) - Binary Metric

**What is Hit Rate?**
- Measures whether the true item appears in the top-K recommendations
- Binary metric: 1 if found, 0 if not found
- Simple and intuitive: "Did we recommend the right item?"

**Example:**
```
True item: Movie #42
Top-10 recommendations: [15, 23, 42, 7, 89, 12, 56, 3, 91, 8]
                        ↑
                    Found at position 2!
Hit Rate = 1 (item is in top-10)
```

**Why Hit Rate?**
- Users typically only see top-K recommendations
- If true item is in top-K, recommendation is successful
- Easy to interpret: "X% of test cases had correct item in top-K"

**Limitations:**
- Doesn't consider position (item at position 1 vs position 10 both get 1)
- Binary: doesn't measure how good the ranking is

---

#### 6.2 NDCG (Normalized Discounted Cumulative Gain) - Ranking Metric

**What is NDCG?**
- Measures ranking quality, not just presence
- Gives more weight to items ranked higher
- Uses logarithmic discounting (relevance decreases with position)

**NDCG Formula:**
```
NDCG = 1 / log2(position + 2)
```

**Position vs NDCG Score:**
| Position | NDCG Score | Meaning |
|----------|-----------|---------|
| 0 (top)  | 1.000     | Perfect! Item ranked #1 |
| 1        | 0.631     | Item ranked #2 |
| 2        | 0.500     | Item ranked #3 |
| 3        | 0.431     | Item ranked #4 |
| 4        | 0.387     | Item ranked #5 |
| 9        | 0.289     | Item ranked #10 |

**Why Logarithmic Discounting?**
- Position 0 → 1: Big difference (user sees it first)
- Position 9 → 10: Small difference (both far down)
- Logarithmic function captures this diminishing importance

**Example:**
```
True item: Movie #42
Top-10 recommendations: [15, 23, 42, 7, 89, 12, 56, 3, 91, 8]
                        ↑
                    Found at position 2!
NDCG = 1 / log2(2 + 2) = 1 / log2(4) = 1 / 2 = 0.5
```

**Why NDCG?**
- Considers position: Better ranking = Higher score
- More informative than Hit Rate
- Standard metric in information retrieval and recommendation systems

**Limitations:**
- More complex than Hit Rate
- Requires understanding of logarithmic discounting

---

#### 6.3 Evaluation Process

**How Evaluation Works:**

1. **For each test case** (1 positive + 99 negatives):
   ```
   User: 123
   Items: [42 (positive), 1, 5, 7, 9, ... (99 negatives)]
   ```

2. **Get predictions**:
   ```
   Model scores: [0.8, 0.3, 0.2, 0.1, 0.05, ...]
   Item 42 gets score 0.8 (highest!)
   ```

3. **Select top-K** (e.g., K=10):
   ```
   Top-10 items: [42, 1, 5, 7, 9, 12, 15, 18, 20, 23]
   ```

4. **Calculate metrics**:
   - Hit Rate: Is 42 in top-10? Yes → HR = 1
   - NDCG: Position of 42? Position 0 → NDCG = 1.0

5. **Average across all test cases**:
   ```
   Mean HR = (1 + 0 + 1 + 1 + ...) / N
   Mean NDCG = (1.0 + 0.0 + 0.5 + 0.63 + ...) / N
   ```

**Test Data Structure:**
- Each batch: 100 items (1 positive + 99 negatives)
- Model should rank the positive item higher than negatives
- We measure if positive is in top-K

**Why 99 Negatives?**
- Simulates real-world scenario: recommend 1 from 100 candidates
- Standard evaluation protocol (used in research papers)
- Makes evaluation realistic and challenging

---

#### 6.4 Understanding the Results

**Good Performance:**
- HR@10 > 0.6: 60% of test cases have true item in top-10
- NDCG@10 > 0.4: Good ranking quality on average

**Excellent Performance:**
- HR@10 > 0.7: 70% success rate
- NDCG@10 > 0.5: Very good ranking quality

**What to Expect:**
- Random baseline: HR@10 ≈ 0.1 (10% chance)
- Good model: HR@10 ≈ 0.6-0.7
- State-of-the-art: HR@10 > 0.7

**Interpreting Results:**
- **HR higher than NDCG**: Model finds items but doesn't rank them well
- **NDCG close to HR**: Model ranks items well (good positions)
- **Both low**: Model needs more training or better architecture

---

**✅ Step 6 Complete!**

We now have:
- Hit Rate metric for binary evaluation
- NDCG metric for ranking quality
- Complete evaluation function ready to use

---

## Step 7: Training Loop

In this step, we'll implement the complete training process:
1. **Loss Function**: Binary Cross-Entropy with Logits (for binary classification)
2. **Optimizer**: Adam optimizer (adaptive learning rate)
3. **Training Loop**: Iterate through epochs, train on batches, evaluate periodically
4. **Model Saving**: Save the best model based on validation performance

This is where the model learns to make good recommendations!

Let's implement this step by step.


In [13]:
# ============================================================================
# STEP 7: TRAINING LOOP
# ============================================================================

"""
This step implements the complete training process for the NCF model.
"""

# ============================================================================
# 7.1 SETUP LOSS FUNCTION AND OPTIMIZER
# ============================================================================

print("=" * 70)
print("STEP 7.1: Setting Up Loss Function and Optimizer")
print("=" * 70)

# Loss Function: Binary Cross-Entropy with Logits
# This combines sigmoid activation + binary cross-entropy loss
# More numerically stable than applying sigmoid separately
# 
# Why BCEWithLogitsLoss?
# - Our task: Predict if user will like item (binary: 1 or 0)
# - Model outputs raw scores (logits), not probabilities
# - BCEWithLogitsLoss applies sigmoid internally and computes loss
# - More stable than: sigmoid(output) then BCE(sigmoid_output, label)

loss_function = nn.BCEWithLogitsLoss()
print("✓ Loss function: BCEWithLogitsLoss")
print("  - For binary classification (like/dislike)")
print("  - Combines sigmoid + cross-entropy for stability")

# Optimizer: Adam (Adaptive Moment Estimation)
# Adam is an adaptive learning rate optimizer that:
# - Adjusts learning rate per parameter
# - Uses momentum (moving average of gradients)
# - Works well for most deep learning tasks
# - Better than SGD for this problem

optimizer = optim.Adam(ncf_model.parameters(), lr=learning_rate)
print(f"\n✓ Optimizer: Adam")
print(f"  - Learning rate: {learning_rate}")
print(f"  - Adaptive: adjusts learning rate automatically")

# Count trainable parameters
total_params = sum(p.numel() for p in ncf_model.parameters() if p.requires_grad)
print(f"\n✓ Model ready for training")
print(f"  - Trainable parameters: {total_params:,}")


STEP 7.1: Setting Up Loss Function and Optimizer
✓ Loss function: BCEWithLogitsLoss
  - For binary classification (like/dislike)
  - Combines sigmoid + cross-entropy for stability

✓ Optimizer: Adam
  - Learning rate: 0.001
  - Adaptive: adjusts learning rate automatically

✓ Model ready for training
  - Trainable parameters: 1,574,657


In [14]:
# ============================================================================
# 7.2 TRAINING LOOP
# ============================================================================

print("\n" + "=" * 70)
print("STEP 7.2: Starting Training")
print("=" * 70)
print(f"Training for {epochs} epochs...")
print(f"Model: {model_name}")
print(f"Device: {device}")
print("=" * 70)

# Track best performance
best_hr = 0.0
best_ndcg = 0.0
best_epoch = 0
training_history = {
    'epoch': [],
    'hr': [],
    'ndcg': [],
    'time': []
}

# Training loop
for epoch in range(epochs):
    # ========================================================================
    # TRAINING PHASE
    # ========================================================================
    
    # Set model to training mode
    # This enables dropout and other training-specific behaviors
    ncf_model.train()
    
    # Start timer for this epoch
    epoch_start_time = time.time()
    
    # Generate negative samples for this epoch
    # Important: We generate fresh negatives each epoch for better learning
    print(f"\nEpoch {epoch+1}/{epochs}")
    print("-" * 70)
    train_dataset.ng_sample()
    
    # Track loss for this epoch
    epoch_loss = 0.0
    num_batches = 0
    
    # Iterate through training batches
    for batch_idx, (user, item, label) in enumerate(train_loader):
        # Move data to device (GPU or CPU)
        if device == 'cuda' and torch.cuda.is_available():
            user = user.cuda()
            item = item.cuda()
            label = label.float().cuda()
        else:
            user = user
            item = item
            label = label.float()
        
        # ================================================================
        # FORWARD PASS
        # ================================================================
        # Clear gradients from previous iteration
        optimizer.zero_grad()
        
        # Get model predictions (raw scores/logits)
        prediction = ncf_model(user, item)  # [batch_size]
        
        # ================================================================
        # COMPUTE LOSS
        # ================================================================
        # Compare predictions with true labels (1 for positive, 0 for negative)
        loss = loss_function(prediction, label)
        
        # ================================================================
        # BACKWARD PASS
        # ================================================================
        # Compute gradients
        loss.backward()
        
        # Update model weights
        optimizer.step()
        
        # Track loss
        epoch_loss += loss.item()
        num_batches += 1
        
        # Print progress every 100 batches
        if (batch_idx + 1) % 100 == 0:
            avg_loss = epoch_loss / num_batches
            print(f"  Batch {batch_idx+1}/{len(train_loader)} - Loss: {avg_loss:.4f}")
    
    # Calculate average loss for this epoch
    avg_loss = epoch_loss / num_batches if num_batches > 0 else 0.0
    
    # ========================================================================
    # EVALUATION PHASE
    # ========================================================================
    
    # Set model to evaluation mode
    # This disables dropout and uses deterministic behavior
    ncf_model.eval()
    
    # Evaluate on test set
    print("  Evaluating on test set...")
    HR, NDCG = evaluate_metrics(ncf_model, test_loader, top_k, device)
    
    # Calculate elapsed time
    elapsed_time = time.time() - epoch_start_time
    time_str = time.strftime("%H:%M:%S", time.gmtime(elapsed_time))
    
    # Store history
    training_history['epoch'].append(epoch + 1)
    training_history['hr'].append(HR)
    training_history['ndcg'].append(NDCG)
    training_history['time'].append(elapsed_time)
    
    # Print epoch results
    print(f"  Time: {time_str}")
    print(f"  Loss: {avg_loss:.4f}")
    print(f"  HR@{top_k}: {HR:.4f}")
    print(f"  NDCG@{top_k}: {NDCG:.4f}")
    
    # ========================================================================
    # SAVE BEST MODEL
    # ========================================================================
    
    # Check if this is the best model so far
    if HR > best_hr:
        best_hr = HR
        best_ndcg = NDCG
        best_epoch = epoch + 1
        
        print(f"  ✓ New best model! (HR@{top_k}: {HR:.4f})")
        
        # Save model if enabled
        if save_model:
            if not os.path.exists(model_path):
                os.makedirs(model_path)
            
            model_filename = os.path.join(model_path, f'{model_name}.pth')
            torch.save(ncf_model, model_filename)
            print(f"  ✓ Model saved to {model_filename}")
    else:
        print(f"  (Best: HR@{top_k}: {best_hr:.4f} at epoch {best_epoch})")
    
    print("-" * 70)

# ========================================================================
# TRAINING COMPLETE
# ========================================================================

print("\n" + "=" * 70)
print("TRAINING COMPLETE!")
print("=" * 70)
print(f"Best model at epoch {best_epoch}:")
print(f"  HR@{top_k}: {best_hr:.4f}")
print(f"  NDCG@{top_k}: {best_ndcg:.4f}")
print("=" * 70)

# Store the NeuMF-end model for later comparison
ncf_model_neumf_end = ncf_model
best_hr_neumf_end = best_hr
best_ndcg_neumf_end = best_ndcg



STEP 7.2: Starting Training
Training for 20 epochs...
Model: NeuMF-end
Device: cpu

Epoch 1/20
----------------------------------------------------------------------
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
  Batch 100/8989 - Loss: 0.4856
  Batch 200/8989 - Loss: 0.4368
  Batch 300/8989 - Loss: 0.4145
  Batch 400/8989 - Loss: 0.4014
  Batch 500/8989 - Loss: 0.3922
  Batch 600/8989 - Loss: 0.3861
  Batch 700/8989 - Loss: 0.3830
  Batch 800/8989 - Loss: 0.3795
  Batch 900/8989 - Loss: 0.3767
  Batch 1000/8989 - Loss: 0.3740
  Batch 1100/8989 - Loss: 0.3719
  Batch 1200/8989 - Loss: 0.3707
  Batch 1300/8989 - Loss: 0.3699
  Batch 1400/8989 - Loss: 0.3687
  Batch 1500/8989 - Loss: 0.3673
  Batch 1600/8989 - Loss: 0.3664
  Batch 1700/8989 - Loss: 0.3652
  Batch 1800/8989 - Loss: 0.3641
  Batch 1900/8989 - Loss: 0.3631
  Batch 2000/8989 - Loss: 0.3626
  Batch 2100/8989 - Loss: 0.3622
  Batch 22

---

## Step 7.5: Train Models Separately (GMF, MLP, NeuMF-pre)

In this step, we'll train each model architecture separately:
1. **Train GMF model** - Generalized Matrix Factorization only
2. **Train MLP model** - Multi-Layer Perceptron only  
3. **Train NeuMF-pre** - Neural Matrix Factorization using pre-trained GMF and MLP weights

This approach (NeuMF-pre) typically gives the best performance as it leverages pre-trained components.


In [15]:
# ============================================================================
# STEP 7.5: TRAIN MODELS SEPARATELY (GMF, MLP, NeuMF-pre)
# ============================================================================

print("=" * 70)
print("STEP 7.5: Training Models Separately")
print("=" * 70)
print("This will train:")
print("  1. GMF model (Generalized Matrix Factorization)")
print("  2. MLP model (Multi-Layer Perceptron)")
print("  3. NeuMF-pre model (using pre-trained GMF and MLP weights)")
print("=" * 70)

# Store trained models
trained_models = {}

# ============================================================================
# 7.5.1 TRAIN GMF MODEL
# ============================================================================

print("\n" + "=" * 70)
print("STEP 7.5.1: Training GMF Model")
print("=" * 70)

# Create GMF model
gmf_model = NCF(
    user_num=user_num,
    item_num=item_num,
    factor_num=factor_num,
    num_layers=num_layers,
    dropout=dropout_rate,
    model_name='GMF',
    GMF_model=None,
    MLP_model=None
)

if torch.cuda.is_available():
    gmf_model = gmf_model.cuda()

# Setup optimizer and loss
gmf_optimizer = optim.Adam(gmf_model.parameters(), lr=learning_rate)
gmf_loss_function = nn.BCEWithLogitsLoss()

# Training loop for GMF
print(f"Training GMF for {epochs} epochs...")
best_hr_gmf = 0.0
best_ndcg_gmf = 0.0
best_epoch_gmf = 0

for epoch in range(epochs):
    gmf_model.train()
    epoch_start_time = time.time()
    train_dataset.ng_sample()
    
    epoch_loss = 0.0
    num_batches = 0
    
    for batch_idx, (user, item, label) in enumerate(train_loader):
        if device == 'cuda' and torch.cuda.is_available():
            user = user.cuda()
            item = item.cuda()
            label = label.float().cuda()
        else:
            user = user
            item = item
            label = label.float()
        
        gmf_optimizer.zero_grad()
        prediction = gmf_model(user, item)
        loss = gmf_loss_function(prediction, label)
        loss.backward()
        gmf_optimizer.step()
        
        epoch_loss += loss.item()
        num_batches += 1
    
    avg_loss = epoch_loss / num_batches if num_batches > 0 else 0.0
    
    # Evaluate
    gmf_model.eval()
    HR, NDCG = evaluate_metrics(gmf_model, test_loader, top_k, device)
    
    elapsed_time = time.time() - epoch_start_time
    time_str = time.strftime("%H:%M:%S", time.gmtime(elapsed_time))
    
    print(f"Epoch {epoch+1}/{epochs} - Time: {time_str} - Loss: {avg_loss:.4f} - HR@{top_k}: {HR:.4f} - NDCG@{top_k}: {NDCG:.4f}")
    
    if HR > best_hr_gmf:
        best_hr_gmf = HR
        best_ndcg_gmf = NDCG
        best_epoch_gmf = epoch + 1
        if save_model:
            torch.save(gmf_model, GMF_model_path)
            print(f"  ✓ Saved best GMF model (HR@{top_k}: {HR:.4f})")

print(f"\n✓ GMF Training Complete!")
print(f"  Best epoch: {best_epoch_gmf}")
print(f"  Best HR@{top_k}: {best_hr_gmf:.4f}")
print(f"  Best NDCG@{top_k}: {best_ndcg_gmf:.4f}")

trained_models['GMF'] = {
    'model': gmf_model,
    'hr': best_hr_gmf,
    'ndcg': best_ndcg_gmf,
    'epoch': best_epoch_gmf
}


STEP 7.5: Training Models Separately
This will train:
  1. GMF model (Generalized Matrix Factorization)
  2. MLP model (Multi-Layer Perceptron)
  3. NeuMF-pre model (using pre-trained GMF and MLP weights)

STEP 7.5.1: Training GMF Model
Training GMF for 20 epochs...
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 1/20 - Time: 00:00:20 - Loss: 0.3577 - HR@10: 0.6525 - NDCG@10: 0.3838
  ✓ Saved best GMF model (HR@10: 0.6525)
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 2/20 - Time: 00:00:20 - Loss: 0.2979 - HR@10: 0.7042 - NDCG@10: 0.4259
  ✓ Saved best GMF model (HR@10: 0.7042)
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 3/20 - Time: 00:00:19 - Loss: 0.2781 - HR@10: 0.7249 - NDCG@10: 0.4432
  ✓

In [16]:
# ============================================================================
# 7.5.2 TRAIN MLP MODEL
# ============================================================================

print("\n" + "=" * 70)
print("STEP 7.5.2: Training MLP Model")
print("=" * 70)

# Create MLP model
mlp_model = NCF(
    user_num=user_num,
    item_num=item_num,
    factor_num=factor_num,
    num_layers=num_layers,
    dropout=dropout_rate,
    model_name='MLP',
    GMF_model=None,
    MLP_model=None
)

if torch.cuda.is_available():
    mlp_model = mlp_model.cuda()

# Setup optimizer and loss
mlp_optimizer = optim.Adam(mlp_model.parameters(), lr=learning_rate)
mlp_loss_function = nn.BCEWithLogitsLoss()

# Training loop for MLP
print(f"Training MLP for {epochs} epochs...")
best_hr_mlp = 0.0
best_ndcg_mlp = 0.0
best_epoch_mlp = 0

for epoch in range(epochs):
    mlp_model.train()
    epoch_start_time = time.time()
    train_dataset.ng_sample()
    
    epoch_loss = 0.0
    num_batches = 0
    
    for batch_idx, (user, item, label) in enumerate(train_loader):
        if device == 'cuda' and torch.cuda.is_available():
            user = user.cuda()
            item = item.cuda()
            label = label.float().cuda()
        else:
            user = user
            item = item
            label = label.float()
        
        mlp_optimizer.zero_grad()
        prediction = mlp_model(user, item)
        loss = mlp_loss_function(prediction, label)
        loss.backward()
        mlp_optimizer.step()
        
        epoch_loss += loss.item()
        num_batches += 1
    
    avg_loss = epoch_loss / num_batches if num_batches > 0 else 0.0
    
    # Evaluate
    mlp_model.eval()
    HR, NDCG = evaluate_metrics(mlp_model, test_loader, top_k, device)
    
    elapsed_time = time.time() - epoch_start_time
    time_str = time.strftime("%H:%M:%S", time.gmtime(elapsed_time))
    
    print(f"Epoch {epoch+1}/{epochs} - Time: {time_str} - Loss: {avg_loss:.4f} - HR@{top_k}: {HR:.4f} - NDCG@{top_k}: {NDCG:.4f}")
    
    if HR > best_hr_mlp:
        best_hr_mlp = HR
        best_ndcg_mlp = NDCG
        best_epoch_mlp = epoch + 1
        if save_model:
            torch.save(mlp_model, MLP_model_path)
            print(f"  ✓ Saved best MLP model (HR@{top_k}: {HR:.4f})")

print(f"\n✓ MLP Training Complete!")
print(f"  Best epoch: {best_epoch_mlp}")
print(f"  Best HR@{top_k}: {best_hr_mlp:.4f}")
print(f"  Best NDCG@{top_k}: {best_ndcg_mlp:.4f}")

trained_models['MLP'] = {
    'model': mlp_model,
    'hr': best_hr_mlp,
    'ndcg': best_ndcg_mlp,
    'epoch': best_epoch_mlp
}



STEP 7.5.2: Training MLP Model
Training MLP for 20 epochs...
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 1/20 - Time: 00:00:33 - Loss: 0.3491 - HR@10: 0.5890 - NDCG@10: 0.3364
  ✓ Saved best MLP model (HR@10: 0.5890)
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 2/20 - Time: 00:00:32 - Loss: 0.3182 - HR@10: 0.6518 - NDCG@10: 0.3826
  ✓ Saved best MLP model (HR@10: 0.6518)
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 3/20 - Time: 00:00:33 - Loss: 0.2971 - HR@10: 0.6923 - NDCG@10: 0.4143
  ✓ Saved best MLP model (HR@10: 0.6923)
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 4/20 - Time: 00:00:32 

In [17]:
# ============================================================================
# 7.5.3 TRAIN NeuMF-pre MODEL (Using Pre-trained GMF and MLP)
# ============================================================================

print("\n" + "=" * 70)
print("STEP 7.5.3: Training NeuMF-pre Model")
print("=" * 70)
print("Creating NeuMF model with pre-trained GMF and MLP weights...")

# Create NeuMF-pre model using pre-trained GMF and MLP
neumf_pre_model = NCF(
    user_num=user_num,
    item_num=item_num,
    factor_num=factor_num,
    num_layers=num_layers,
    dropout=dropout_rate,
    model_name='NeuMF-pre',
    GMF_model=gmf_model,
    MLP_model=mlp_model
)

if torch.cuda.is_available():
    neumf_pre_model = neumf_pre_model.cuda()

# Setup optimizer (SGD is typically used for NeuMF-pre)
neumf_pre_optimizer = optim.SGD(neumf_pre_model.parameters(), lr=learning_rate)
neumf_pre_loss_function = nn.BCEWithLogitsLoss()

# Training loop for NeuMF-pre
print(f"Training NeuMF-pre for {epochs} epochs...")
best_hr_neumf_pre = 0.0
best_ndcg_neumf_pre = 0.0
best_epoch_neumf_pre = 0

for epoch in range(epochs):
    neumf_pre_model.train()
    epoch_start_time = time.time()
    train_dataset.ng_sample()
    
    epoch_loss = 0.0
    num_batches = 0
    
    for batch_idx, (user, item, label) in enumerate(train_loader):
        if device == 'cuda' and torch.cuda.is_available():
            user = user.cuda()
            item = item.cuda()
            label = label.float().cuda()
        else:
            user = user
            item = item
            label = label.float()
        
        neumf_pre_optimizer.zero_grad()
        prediction = neumf_pre_model(user, item)
        loss = neumf_pre_loss_function(prediction, label)
        loss.backward()
        neumf_pre_optimizer.step()
        
        epoch_loss += loss.item()
        num_batches += 1
    
    avg_loss = epoch_loss / num_batches if num_batches > 0 else 0.0
    
    # Evaluate
    neumf_pre_model.eval()
    HR, NDCG = evaluate_metrics(neumf_pre_model, test_loader, top_k, device)
    
    elapsed_time = time.time() - epoch_start_time
    time_str = time.strftime("%H:%M:%S", time.gmtime(elapsed_time))
    
    print(f"Epoch {epoch+1}/{epochs} - Time: {time_str} - Loss: {avg_loss:.4f} - HR@{top_k}: {HR:.4f} - NDCG@{top_k}: {NDCG:.4f}")
    
    if HR > best_hr_neumf_pre:
        best_hr_neumf_pre = HR
        best_ndcg_neumf_pre = NDCG
        best_epoch_neumf_pre = epoch + 1
        if save_model:
            torch.save(neumf_pre_model, NeuMF_model_path)
            print(f"  ✓ Saved best NeuMF-pre model (HR@{top_k}: {HR:.4f})")

print(f"\n✓ NeuMF-pre Training Complete!")
print(f"  Best epoch: {best_epoch_neumf_pre}")
print(f"  Best HR@{top_k}: {best_hr_neumf_pre:.4f}")
print(f"  Best NDCG@{top_k}: {best_ndcg_neumf_pre:.4f}")

trained_models['NeuMF-pre'] = {
    'model': neumf_pre_model,
    'hr': best_hr_neumf_pre,
    'ndcg': best_ndcg_neumf_pre,
    'epoch': best_epoch_neumf_pre
}

# ============================================================================
# COMPARISON OF ALL MODELS
# ============================================================================

print("\n" + "=" * 70)
print("MODEL COMPARISON SUMMARY")
print("=" * 70)
print(f"{'Model':<15} {'HR@{top_k}':<12} {'NDCG@{top_k}':<12} {'Best Epoch':<12}")
print("-" * 70)
print(f"{'GMF':<15} {best_hr_gmf:<12.4f} {best_ndcg_gmf:<12.4f} {best_epoch_gmf:<12}")
print(f"{'MLP':<15} {best_hr_mlp:<12.4f} {best_ndcg_mlp:<12.4f} {best_epoch_mlp:<12}")
print(f"{'NeuMF-end':<15} {best_hr_neumf_end:<12.4f} {best_ndcg_neumf_end:<12.4f} {best_epoch:<12}")
print(f"{'NeuMF-pre':<15} {best_hr_neumf_pre:<12.4f} {best_ndcg_neumf_pre:<12.4f} {best_epoch_neumf_pre:<12}")
print("=" * 70)

# Find best model
all_results = [
    ('GMF', best_hr_gmf, best_ndcg_gmf),
    ('MLP', best_hr_mlp, best_ndcg_mlp),
    ('NeuMF-end', best_hr_neumf_end, best_ndcg_neumf_end),
    ('NeuMF-pre', best_hr_neumf_pre, best_ndcg_neumf_pre)
]
best_model_name, best_hr_overall, best_ndcg_overall = max(all_results, key=lambda x: x[1])

print(f"\n🏆 Best Model: {best_model_name}")
print(f"   HR@{top_k}: {best_hr_overall:.4f}")
print(f"   NDCG@{top_k}: {best_ndcg_overall:.4f}")

# Set the best model as the main model for recommendations
if best_model_name == 'GMF':
    ncf_model = gmf_model
elif best_model_name == 'MLP':
    ncf_model = mlp_model
elif best_model_name == 'NeuMF-pre':
    ncf_model = neumf_pre_model
else:
    ncf_model = ncf_model_neumf_end

print(f"\n✓ Using {best_model_name} model for recommendations")
print("=" * 70)



STEP 7.5.3: Training NeuMF-pre Model
Creating NeuMF model with pre-trained GMF and MLP weights...
Training NeuMF-pre for 20 epochs...
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 1/20 - Time: 00:00:28 - Loss: 0.1674 - HR@10: 0.7554 - NDCG@10: 0.4782
  ✓ Saved best NeuMF-pre model (HR@10: 0.7554)
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 2/20 - Time: 00:00:28 - Loss: 0.1663 - HR@10: 0.7531 - NDCG@10: 0.4764
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 3/20 - Time: 00:00:28 - Loss: 0.1648 - HR@10: 0.7517 - NDCG@10: 0.4752
Generating 4 negative samples per positive pair...
✓ Generated 1840900 negative samples
  - Total samples (positives + negatives): 2301125
Epoch 4/20 - Time: 00:00:28 - L

### Explanation of Step 7.5:

**Why Train Separately?**

1. **GMF (Generalized Matrix Factorization)**:
   - Simple linear model
   - Fast to train
   - Good baseline performance
   - Captures linear user-item interactions

2. **MLP (Multi-Layer Perceptron)**:
   - Deep non-linear model
   - Can learn complex patterns
   - Complements GMF's linear approach

3. **NeuMF-pre (Pre-trained Neural Matrix Factorization)**:
   - Combines pre-trained GMF and MLP
   - Initializes with learned embeddings from both models
   - Typically achieves best performance
   - Fine-tunes the combined model

**Training Strategy:**
- Train GMF and MLP independently
- Use their learned embeddings to initialize NeuMF
- Fine-tune NeuMF with SGD (more stable than Adam for pre-trained models)
- This gives better performance than training NeuMF from scratch

**Model Comparison:**
- GMF: Fast, simple, good baseline
- MLP: Complex patterns, non-linear
- NeuMF-end: Trained from scratch, good balance
- NeuMF-pre: Best performance, uses pre-trained components

---

**✅ Step 7.5 Complete!**

All models have been trained separately. The best model is now available for recommendations.


### Detailed Explanation of Step 7:

#### 7.1 Loss Function and Optimizer

**Binary Cross-Entropy with Logits Loss (BCEWithLogitsLoss):**

**What it does:**
- Combines sigmoid activation + binary cross-entropy loss
- More numerically stable than applying sigmoid separately
- Formula: `loss = -[y*log(σ(x)) + (1-y)*log(1-σ(x))]`
  - Where σ(x) = sigmoid(x), y = true label (0 or 1), x = model output

**Why this loss?**
- Our task: Predict if user will like item (binary classification)
- Labels: 1 (positive interaction) or 0 (negative interaction)
- Model outputs: Raw scores (logits), not probabilities
- BCEWithLogitsLoss handles the conversion internally

**Numerical Stability:**
- Direct sigmoid can cause overflow/underflow
- BCEWithLogitsLoss uses log-sum-exp trick for stability
- Prevents NaN/inf values during training

**Adam Optimizer:**

**What is Adam?**
- Adaptive Moment Estimation
- Combines benefits of:
  - Momentum: Uses moving average of gradients (smoother updates)
  - RMSprop: Adapts learning rate per parameter
  - Bias correction: Accounts for initialization bias

**Why Adam?**
- Works well out-of-the-box (default hyperparameters)
- Adapts learning rate automatically
- Faster convergence than SGD for most problems
- Good for recommendation systems

**Learning Rate:**
- 0.001 is a safe default for Adam
- Too high: Training unstable, loss might explode
- Too low: Training very slow, might not converge

---

#### 7.2 Training Loop - Step by Step

**Epoch Structure:**

1. **Set Training Mode**:
   ```python
   model.train()  # Enables dropout, batch norm training mode
   ```

2. **Generate Negative Samples**:
   ```python
   train_dataset.ng_sample()  # Fresh negatives each epoch
   ```
   - Important: New negatives each epoch improves learning
   - Prevents overfitting to specific negative samples

3. **Iterate Through Batches**:
   ```python
   for user, item, label in train_loader:
       # Process batch
   ```

4. **Forward Pass**:
   ```python
   prediction = model(user, item)  # Get model predictions
   ```
   - Model outputs raw scores (logits)
   - Higher score = more likely user will like item

5. **Compute Loss**:
   ```python
   loss = loss_function(prediction, label)
   ```
   - Compares predictions with true labels
   - Measures how wrong the model is

6. **Backward Pass**:
   ```python
   loss.backward()  # Compute gradients
   optimizer.step()  # Update weights
   ```
   - Gradients tell us how to adjust weights
   - Optimizer updates weights to reduce loss

7. **Evaluation**:
   ```python
   model.eval()  # Disable dropout
   HR, NDCG = evaluate_metrics(model, test_loader, top_k)
   ```
   - Test model on held-out test set
   - Measure performance with Hit Rate and NDCG

8. **Save Best Model**:
   ```python
   if HR > best_hr:
       torch.save(model, 'best_model.pth')
   ```
   - Save model with best validation performance
   - Prevents losing good models if training degrades

---

#### 7.3 Understanding Training Progress

**What to Watch:**

1. **Loss Decreasing**:
   - Should decrease over epochs
   - If increases: learning rate too high or model unstable
   - If plateaus: model converged or needs more capacity

2. **HR and NDCG Increasing**:
   - Should improve over epochs
   - Early epochs: rapid improvement
   - Later epochs: slower improvement (diminishing returns)

3. **Best Model Tracking**:
   - Model might overfit (train loss decreases, test HR decreases)
   - We save best model based on test HR
   - This prevents using overfitted model

**Typical Training Curve:**
```
Epoch 1:  HR: 0.200, NDCG: 0.100  (random)
Epoch 5:  HR: 0.500, NDCG: 0.300  (learning)
Epoch 10: HR: 0.650, NDCG: 0.420  (improving)
Epoch 15: HR: 0.680, NDCG: 0.440  (slowing)
Epoch 20: HR: 0.690, NDCG: 0.445  (converged)
```

**When to Stop:**
- If HR stops improving for several epochs
- If test HR starts decreasing (overfitting)
- After specified number of epochs

---

#### 7.4 Model Saving

**Why Save Models?**
- Training takes time (minutes to hours)
- Best model might not be the last epoch
- Allows loading and using model later
- Enables comparison of different runs

**What Gets Saved:**
- All model parameters (weights and biases)
- Model architecture (can reconstruct model)
- Optimizer state (optional, for resuming training)

**Loading Saved Model:**
```python
model = torch.load('best_model.pth')
model.eval()  # Set to evaluation mode
```

---

#### 7.5 Training Tips

**If Training is Slow:**
- Reduce batch size (if memory allows)
- Use GPU if available
- Reduce number of epochs (if converged early)

**If Loss Not Decreasing:**
- Check learning rate (might be too high/low)
- Check data loading (make sure data is correct)
- Check model architecture (might have bugs)

**If Overfitting (train good, test bad):**
- Increase dropout rate
- Reduce model capacity (fewer layers/embeddings)
- Get more training data
- Use early stopping

**If Underfitting (both train and test bad):**
- Increase model capacity
- Train for more epochs
- Reduce dropout
- Check if learning rate is appropriate

---

**✅ Step 7 Complete!**

We now have:
- Complete training loop implemented
- Model training and evaluation
- Best model saving
- Progress tracking

---

## Step 8: Using the Trained Model for Recommendations

Now that we have a trained model, let's learn how to use it to make recommendations! This step shows:
1. How to load a saved model
2. How to get recommendations for a user
3. How to predict user-item interaction scores
4. Practical examples

This is where the model becomes useful for real-world applications!


In [18]:
# ============================================================================
# STEP 8: USING THE TRAINED MODEL FOR RECOMMENDATIONS
# ============================================================================

"""
This step demonstrates how to use the trained NCF model to make recommendations.
"""

# ============================================================================
# 8.1 LOAD SAVED MODEL (Optional - if you want to load from disk)
# ============================================================================

print("=" * 70)
print("STEP 8: Using Trained Model for Recommendations")
print("=" * 70)

# If you saved a model and want to load it later, use this:
# model_path_to_load = os.path.join(model_path, f'{model_name}.pth')
# if os.path.exists(model_path_to_load):
#     print(f"Loading model from {model_path_to_load}...")
#     ncf_model = torch.load(model_path_to_load)
#     if torch.cuda.is_available():
#         ncf_model = ncf_model.cuda()
#     ncf_model.eval()
#     print("✓ Model loaded successfully!")
# else:
#     print("Using currently trained model...")

# For now, we'll use the model we just trained
print("Using the trained model from Step 7...")
model_path_to_load = os.path.join(model_path, f'{model_name}.pth')
if os.path.exists(model_path_to_load):

    print("✓ Model ready for inference")
    model = torch.load('models/NeuMF-end.pth' , weights_only=False)
    model.eval()  # Important: set to evaluation mode


STEP 8: Using Trained Model for Recommendations
Using the trained model from Step 7...
✓ Model ready for inference


In [19]:
# ============================================================================
# 8.2 GET TOP-K RECOMMENDATIONS FOR A USER
# ============================================================================

def get_top_k_recommendations(model, user_id, item_ids, k=10, device='cpu'):
    """
    Get top-K item recommendations for a given user.
    
    Parameters:
    - model: Trained NCF model
    - user_id: ID of the user (integer)
    - item_ids: List of item IDs to consider (e.g., all items or candidate items)
    - k: Number of recommendations to return
    - device: 'cuda' or 'cpu'
    
    Returns:
    - top_k_items: List of top-K recommended item IDs
    - top_k_scores: List of corresponding prediction scores
    """
    model.eval()  # Set to evaluation mode
    
    # Convert to tensors
    user_tensor = torch.LongTensor([user_id] * len(item_ids))
    item_tensor = torch.LongTensor(item_ids)
    
    # Move to device
    if device == 'cuda' and torch.cuda.is_available():
        user_tensor = user_tensor.cuda()
        item_tensor = item_tensor.cuda()
    
    # Get predictions
    with torch.no_grad():
        scores = model(user_tensor, item_tensor)
        scores = scores.cpu().numpy()
    
    # Get top-K items
    top_k_indices = np.argsort(scores)[::-1][:k]  # Sort descending, take top K
    top_k_items = [item_ids[i] for i in top_k_indices]
    top_k_scores = scores[top_k_indices].tolist()
    
    return top_k_items, top_k_scores

print("\n" + "=" * 70)
print("STEP 8.2: Recommendation Function")
print("=" * 70)
print("✓ get_top_k_recommendations() function defined")
print("  - Takes user ID and candidate items")
print("  - Returns top-K recommendations with scores")



STEP 8.2: Recommendation Function
✓ get_top_k_recommendations() function defined
  - Takes user ID and candidate items
  - Returns top-K recommendations with scores


In [20]:
# ============================================================================
# 8.3 PREDICT USER-ITEM INTERACTION SCORE
# ============================================================================

def predict_interaction_score(model, user_id, item_id, device='cpu'):
    """
    Predict the interaction score for a specific user-item pair.
    
    Parameters:
    - model: Trained NCF model
    - user_id: ID of the user (integer)
    - item_id: ID of the item (integer)
    - device: 'cuda' or 'cpu'
    
    Returns:
    - score: Prediction score (higher = more likely user will like item)
    """
    model.eval()
    
    # Convert to tensors
    user_tensor = torch.LongTensor([user_id])
    item_tensor = torch.LongTensor([item_id])
    
    # Move to device
    if device == 'cuda' and torch.cuda.is_available():
        user_tensor = user_tensor.cuda()
        item_tensor = item_tensor.cuda()
    
    # Get prediction
    with torch.no_grad():
        score = model(user_tensor, item_tensor)
        score = score.cpu().item()
    
    return score

print("\n" + "=" * 70)
print("STEP 8.3: Prediction Function")
print("=" * 70)
print("✓ predict_interaction_score() function defined")
print("  - Predicts score for a single user-item pair")
print("  - Returns a single score value")



STEP 8.3: Prediction Function
✓ predict_interaction_score() function defined
  - Predicts score for a single user-item pair
  - Returns a single score value


In [21]:
# ============================================================================
# 8.4 EXAMPLE: GET RECOMMENDATIONS FOR A USER
# ============================================================================

print("\n" + "=" * 70)
print("STEP 8.4: Example Usage")
print("=" * 70)

# Example: Get recommendations for user 0
example_user_id = 0
print(f"\nGetting top-{top_k} recommendations for User {example_user_id}...")

# Get all item IDs (excluding items user already interacted with in training)
# In practice, you might want to filter out items the user has already seen
all_item_ids = list(range(item_num))

# Get top-K recommendations
recommended_items, recommended_scores = get_top_k_recommendations(
    model,
    example_user_id,
    all_item_ids,
    k=top_k,
    device=device
)

print(f"\n✓ Top-{top_k} Recommendations for User {example_user_id}:")
print("-" * 70)
for i, (item_id, score) in enumerate(zip(recommended_items, recommended_scores), 1):
    print(f"  {i}. Item {item_id:5d} - Score: {score:7.4f}")

# Example: Predict score for a specific user-item pair
print(f"\n" + "=" * 70)
print("Example: Predicting interaction score")
print("=" * 70)
example_item_id = recommended_items[0]  # Use the top recommended item
score = predict_interaction_score(model, example_user_id, example_item_id, device=device)
print(f"User {example_user_id} - Item {example_item_id}: Score = {score:.4f}")
print(f"  → Higher score means user is more likely to like this item")



STEP 8.4: Example Usage

Getting top-10 recommendations for User 0...

✓ Top-10 Recommendations for User 0:
----------------------------------------------------------------------
  1. Item   565 - Score:  4.2599
  2. Item   559 - Score:  3.9050
  3. Item    33 - Score:  3.8329
  4. Item   563 - Score:  3.7590
  5. Item  1125 - Score:  3.5868
  6. Item  1113 - Score:  3.5455
  7. Item   344 - Score:  3.4968
  8. Item  1811 - Score:  3.4611
  9. Item   970 - Score:  3.4368
  10. Item   299 - Score:  3.3611

Example: Predicting interaction score
User 0 - Item 565: Score = 4.2599
  → Higher score means user is more likely to like this item


In [22]:
# ============================================================================
# 8.5 FILTER OUT ITEMS USER ALREADY INTERACTED WITH
# ============================================================================

def get_recommendations_excluding_training(user_id, model, train_mat, all_items, k=10, device='cpu'):
    """
    Get recommendations for a user, excluding items they've already interacted with.
    
    This is more realistic - we don't want to recommend items the user already knows about.
    
    Parameters:
    - user_id: ID of the user
    - model: Trained NCF model
    - train_mat: Training interaction matrix (to check what user already interacted with)
    - all_items: List of all item IDs
    - k: Number of recommendations
    - device: 'cuda' or 'cpu'
    
    Returns:
    - top_k_items: Recommended item IDs (excluding training items)
    - top_k_scores: Corresponding scores
    """
    # Filter out items user already interacted with
    candidate_items = [item_id for item_id in all_items 
                       if (user_id, item_id) not in train_mat]
    
    if len(candidate_items) < k:
        print(f"Warning: Only {len(candidate_items)} candidate items available (requested {k})")
        k = len(candidate_items)
    
    # Get recommendations from candidate items
    top_k_items, top_k_scores = get_top_k_recommendations(
        model, user_id, candidate_items, k=k, device=device
    )
    
    return top_k_items, top_k_scores

print("\n" + "=" * 70)
print("STEP 8.5: Filtered Recommendations")
print("=" * 70)

# Example with filtering
example_user_id = 10
print(f"\nGetting filtered recommendations for User {example_user_id}...")
print("(Excluding items user already interacted with in training)")

filtered_items, filtered_scores = get_recommendations_excluding_training(
    example_user_id,
    ncf_model,
    train_mat,
    list(range(item_num)),
    k=top_k,
    device=device
)

print(f"\n✓ Top-{top_k} NEW Recommendations for User {example_user_id}:")
print("-" * 70)
for i, (item_id, score) in enumerate(zip(filtered_items, filtered_scores), 1):
    print(f"  {i}. Item {item_id:5d} - Score: {score:7.4f}")

print("\n" + "=" * 70)
print("✓ Recommendation system ready!")
print("=" * 70)



STEP 8.5: Filtered Recommendations

Getting filtered recommendations for User 10...
(Excluding items user already interacted with in training)

✓ Top-10 NEW Recommendations for User 10:
----------------------------------------------------------------------
  1. Item  2526 - Score:  3.8752
  2. Item  2659 - Score:  3.4374
  3. Item   939 - Score:  3.2481
  4. Item  1325 - Score:  3.1528
  5. Item  1650 - Score:  3.1314
  6. Item   211 - Score:  2.9503
  7. Item  2093 - Score:  2.9223
  8. Item  1499 - Score:  2.9186
  9. Item   278 - Score:  2.8437
  10. Item   324 - Score:  2.7961

✓ Recommendation system ready!


### Detailed Explanation of Step 8:

#### 8.1 Loading a Saved Model

**When to Load:**
- After training, if you saved the model and want to use it later
- When deploying the model in production
- When sharing the model with others

**How to Load:**
```python
model = torch.load('path/to/model.pth')
model.eval()  # Important: set to evaluation mode
```

**Why `model.eval()`?**
- Disables dropout (uses all neurons)
- Uses deterministic behavior
- Required for consistent predictions

---

#### 8.2 Getting Top-K Recommendations

**How It Works:**
1. **Input**: User ID and list of candidate items
2. **Process**: 
   - Get model predictions for all candidate items
   - Sort items by prediction score (descending)
   - Return top-K items
3. **Output**: List of recommended item IDs and their scores

**Example Use Case:**
```python
# Recommend 10 movies for user 123
recommendations, scores = get_top_k_recommendations(
    model, user_id=123, item_ids=all_movie_ids, k=10
)
```

**Performance Considerations:**
- If you have many items (millions), consider:
  - Pre-filtering candidates (e.g., by genre, popularity)
  - Using approximate nearest neighbor search
  - Caching embeddings for faster computation

---

#### 8.3 Predicting Single Interaction Score

**When to Use:**
- Check if a specific user will like a specific item
- Rank a small set of items
- A/B testing different items

**Example:**
```python
score = predict_interaction_score(model, user_id=123, item_id=456)
if score > 0.5:  # Threshold (adjust based on your data)
    print("User will likely like this item")
```

**Interpreting Scores:**
- Higher score = more likely user will like item
- Scores are logits (not probabilities)
- To get probabilities: `prob = torch.sigmoid(torch.tensor(score))`

---

#### 8.4 Filtering Training Items

**Why Filter?**
- Users have already seen/interacted with training items
- We want to recommend NEW items
- More realistic recommendation scenario

**How It Works:**
1. Check training matrix: `(user_id, item_id) in train_mat`
2. Exclude items user already interacted with
3. Get recommendations from remaining items

**Example:**
```python
# Get new recommendations (excluding items user already knows)
recommendations = get_recommendations_excluding_training(
    user_id, model, train_mat, all_items, k=10
)
```

---

#### 8.5 Real-World Usage Tips

**1. Pre-compute Embeddings (Optional):**
```python
# For faster recommendations, pre-compute item embeddings
with torch.no_grad():
    item_embeddings = model.embed_item_GMF(torch.arange(item_num))
# Then use these for faster similarity search
```

**2. Batch Predictions:**
```python
# Predict for multiple users at once (faster)
user_ids = torch.LongTensor([1, 2, 3, 4, 5])
item_ids = torch.LongTensor([10, 20, 30, 40, 50])
scores = model(user_ids, item_ids)  # Batch prediction
```

**3. Cold Start Problem:**
- New users: No interaction history
- Solutions: Use popularity-based recommendations, ask for preferences
- New items: No interaction history
- Solutions: Use content-based features, wait for initial interactions

**4. Evaluation in Production:**
- A/B testing: Compare different models
- Online metrics: Click-through rate, conversion rate
- Offline metrics: HR, NDCG (what we used)

---

**✅ Step 8 Complete!**

We now have:
- Functions to get recommendations
- Functions to predict interaction scores
- Example usage code
- Understanding of how to use the model in practice

---

**🎉 Congratulations! You've completed the full NCF Tutorial!**

**What You've Learned:**
1. ✅ Environment setup and imports
2. ✅ Configuration and hyperparameters
3. ✅ Data downloading and preprocessing
4. ✅ PyTorch Dataset class implementation
5. ✅ NCF model architecture (with detailed visualization)
6. ✅ Evaluation metrics (Hit Rate and NDCG)
7. ✅ Training loop implementation
8. ✅ Using the model for recommendations

**The Complete Pipeline:**
```
Data → Preprocessing → Dataset → Model → Training → Evaluation → Recommendations
```

**Next Steps:**
- Experiment with different hyperparameters
- Try different model architectures (GMF, MLP, NeuMF)
- Test on different datasets
- Deploy for production use
- Explore advanced techniques (attention mechanisms, graph neural networks)

**Thank you for following along! Happy recommending! 🚀**


---

## Step 9: Loading and Using Saved Models

This final step shows how to load the saved models and use them for recommendations. All models (GMF, MLP, NeuMF-end, NeuMF-pre) have been saved and can be loaded independently.


In [23]:
# ============================================================================
# STEP 9: LOADING AND USING SAVED MODELS
# ============================================================================

print("=" * 70)
print("STEP 9: Loading and Using Saved Models")
print("=" * 70)

# ============================================================================
# 9.1 LOAD ALL SAVED MODELS
# ============================================================================

print("\n" + "=" * 70)
print("STEP 9.1: Loading Saved Models")
print("=" * 70)

loaded_models = {}

# Load GMF model
if os.path.exists(GMF_model_path):
    print(f"Loading GMF model from {GMF_model_path}...")
    gmf_loaded = torch.load(GMF_model_path, map_location=device)
    if device == 'cpu':
        gmf_loaded = gmf_loaded.cpu()
    gmf_loaded.eval()
    loaded_models['GMF'] = gmf_loaded
    print("✓ GMF model loaded")
else:
    print(f"⚠ GMF model not found at {GMF_model_path}")

# Load MLP model
if os.path.exists(MLP_model_path):
    print(f"Loading MLP model from {MLP_model_path}...")
    mlp_loaded = torch.load(MLP_model_path, map_location=device)
    if device == 'cpu':
        mlp_loaded = mlp_loaded.cpu()
    mlp_loaded.eval()
    loaded_models['MLP'] = mlp_loaded
    print("✓ MLP model loaded")
else:
    print(f"⚠ MLP model not found at {MLP_model_path}")

# Load NeuMF-end model
neumf_end_path = os.path.join(model_path, 'NeuMF-end.pth')
if os.path.exists(neumf_end_path):
    print(f"Loading NeuMF-end model from {neumf_end_path}...")
    neumf_end_loaded = torch.load(neumf_end_path, map_location=device)
    if device == 'cpu':
        neumf_end_loaded = neumf_end_loaded.cpu()
    neumf_end_loaded.eval()
    loaded_models['NeuMF-end'] = neumf_end_loaded
    print("✓ NeuMF-end model loaded")
else:
    print(f"⚠ NeuMF-end model not found at {neumf_end_path}")

# Load NeuMF-pre model
if os.path.exists(NeuMF_model_path):
    print(f"Loading NeuMF-pre model from {NeuMF_model_path}...")
    neumf_pre_loaded = torch.load(NeuMF_model_path, map_location=device)
    if device == 'cpu':
        neumf_pre_loaded = neumf_pre_loaded.cpu()
    neumf_pre_loaded.eval()
    loaded_models['NeuMF-pre'] = neumf_pre_loaded
    print("✓ NeuMF-pre model loaded")
else:
    print(f"⚠ NeuMF-pre model not found at {NeuMF_model_path}")

print(f"\n✓ Loaded {len(loaded_models)} model(s)")
print(f"  Available models: {list(loaded_models.keys())}")


STEP 9: Loading and Using Saved Models

STEP 9.1: Loading Saved Models
Loading GMF model from ./models/GMF.pth...


UnpicklingError: Weights only load failed. This file can still be loaded, to do so you have two options, do those steps only if you trust the source of the checkpoint. 
	(1) In PyTorch 2.6, we changed the default value of the `weights_only` argument in `torch.load` from `False` to `True`. Re-running `torch.load` with `weights_only` set to `False` will likely succeed, but it can result in arbitrary code execution. Do it only if you got the file from a trusted source.
	(2) Alternatively, to load with `weights_only=True` please check the recommended steps in the following error message.
	WeightsUnpickler error: Unsupported global: GLOBAL __main__.NCF was not an allowed global by default. Please use `torch.serialization.add_safe_globals([__main__.NCF])` or the `torch.serialization.safe_globals([__main__.NCF])` context manager to allowlist this global if you trust this class/function.

Check the documentation of torch.load to learn more about types accepted by default with weights_only https://pytorch.org/docs/stable/generated/torch.load.html.

In [ ]:
# ============================================================================
# 9.2 COMPARE RECOMMENDATIONS FROM ALL MODELS
# ============================================================================

print("\n" + "=" * 70)
print("STEP 9.2: Comparing Recommendations from All Models")
print("=" * 70)

# Example user
example_user_id = 0
print(f"\nGetting recommendations for User {example_user_id} using all models...")

# Get recommendations from each model
all_recommendations = {}

for model_name, model in loaded_models.items():
    if model_name in loaded_models:
        print(f"\n{model_name} recommendations:")
        recommendations, scores = get_top_k_recommendations(
            model, example_user_id, list(range(item_num)), k=top_k, device=device
        )
        all_recommendations[model_name] = (recommendations, scores)
        
        print(f"  Top-{top_k} items:")
        for i, (item_id, score) in enumerate(zip(recommendations[:5], scores[:5]), 1):
            print(f"    {i}. Item {item_id:5d} - Score: {score:7.4f}")
        if top_k > 5:
            print(f"    ... and {top_k - 5} more")

# Compare overlap between models
print("\n" + "=" * 70)
print("Recommendation Overlap Analysis")
print("=" * 70)

if len(all_recommendations) >= 2:
    model_names = list(all_recommendations.keys())
    for i, model1 in enumerate(model_names):
        for model2 in model_names[i+1:]:
            items1 = set(all_recommendations[model1][0])
            items2 = set(all_recommendations[model2][0])
            overlap = len(items1 & items2)
            print(f"{model1} vs {model2}: {overlap}/{top_k} items overlap ({overlap/top_k*100:.1f}%)")
